core/src/subgraph/instance_manager.rs

```diff
@@ -322,16 +328,28 @@ where
         })
         // Process blocks from the stream as long as no restart is needed
         .fold(ctx, move |ctx, block| {
+            let needs_restart_set = needs_restart_set.clone();
+
             process_block(
                 logger.clone(),
                 ctx,
                 block_stream_cancel_handle.clone(),
-                needs_restart.clone(),
                 block,
             )
+            .map(move |(ctx, needs_restart)| {
+                // If new data sources were detected in this block, set the
+                // needs restart flag. This will cause `take_while` to stop
+                // emitting blocks and finish this `fold` loop as well
+                if needs_restart {
+                    needs_restart_set.store(true, Ordering::SeqCst);
+                }
+                ctx
+            })
         })
         // A restart is needed
         .and_then(move |mut ctx| {
+            // Reset the needs restart flag and increase the restart counter
+            needs_restart_reset.store(false, Ordering::SeqCst);
```

Comment:
Must this be reset here? We won't check it again until it's re-instantiated, right?

Reply:
Erm, no! I'm removing it.

Reply:
And it's gone